src/section/section_proof_chain.rs

```diff
@@ -0,0 +1,317 @@
+            }
+        } else {
+            // Root of trust not found
+            return None;
+        }
+
+        let mut current = last_trusted;
+        for block in &self.blocks[block_offset..] {
+            if !validate_next_block(current, block) {
+                return None;
+            }
+
+            current = block.key_info();
+        }
+
+        Some(current)
+    }
+
+    // Verify this proof chain against the given key infos.
+    pub fn check_trust<'a, I>(&'a self, their_key_infos: I) -> TrustStatus<'a>
+    where
+        I: IntoIterator<Item = (&'a Prefix<XorName>, &'a SectionKeyInfo)>,
+    {
+        let first_version = self.version;
+        let (last_prefix, last_version) = self.last_prefix_version();
+        let inclusive_range = first_version..=last_version;
+
+        let mut max_known_version = 0;
+        let mut found_prefix_keys = false;
+
+        for proof_key_info in their_key_infos
+            .into_iter()
+            .filter(|&(pfx, _)| last_prefix.is_compatible(pfx))
+            .map(|(_, info)| info)
+        {
+            max_known_version = std::cmp::max(max_known_version, proof_key_info.version());
+            found_prefix_keys = true;
+
+            if inclusive_range.contains(&proof_key_info.version()) {
+                // We can validate trust with that key: we are done.
+                if let Some(trusted_info) = self.last_trusted_key_info(proof_key_info) {
+                    return TrustStatus::Trusted(trusted_info.key());
+                } else {
+                    return TrustStatus::ProofInvalid;
+                }
+            }
+        }
+
+        if found_prefix_keys && self.version > max_known_version {
+            TrustStatus::ProofTooNew
+        } else {
+            TrustStatus::ProofInvalid
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Hash, Serialize, Deserialize)]
+pub struct SectionProofChain {
+    genesis_key_info: SectionKeyInfo,
+    blocks: Vec<SectionProofBlock>,
+}
+
+impl SectionProofChain {
+    pub fn from_genesis(key_info: SectionKeyInfo) -> Self {
+        Self {
+            genesis_key_info: key_info,
+            blocks: Vec::new(),
+        }
+    }
+
+    pub fn push(&mut self, block: SectionProofBlock) {
+        if !validate_next_block(self.last_key_info(), &block) {
+            log_or_panic!(
+                log::Level::Error,
+                "Invalid next block: {:?} -> {:?}",
+                self.last_key_info(),
+                block
+            );
+            return;
+        }
+
+        self.blocks.push(block)
+    }
+
+    #[cfg(test)]
+    pub fn validate(&self) -> bool {
+        let mut current = &self.genesis_key_info;
+        for block in &self.blocks {
+            if !validate_next_block(current, block) {
+                return false;
+            }
+
+            current = block.key_info();
+        }
+        true
+    }
+
+    pub fn first_key_info(&self) -> &SectionKeyInfo {
```

Comment:
in contrast to the `last_key_info` where we fetch from the blocks, here the first one is always being the genesis ?
this is bit misleading. maybe better have the function name as `genesis_key_info` to be explicit?

Reply:
But the first block IS the genesis block. `blocks.first()` would actually be the second block. The genesis block is stored separately because it doesn't have signature, but it's still part of the same chain.

Reply:
We can treat the genesis block as the network name as that is the ultimate identifier.

Reply:
> The genesis block is stored separately because it doesn't have signature

Ah, ok then.